File: scripts/assets.ts
#!/usr/bin/env bun

import { mkdirSync, readdirSync, copyFileSync, statSync } from "fs";
import { join, basename } from "path";

async function main(): Promise<void> {
  try {
    // Define source and destination paths
    const sourceDir = "assets/fav";
    const releaseDir = "target/dx/hmziq-xyz/release/web/public/assets/fav";

    // Create the destination directory structure if it doesn't exist
    mkdirSync(releaseDir, { recursive: true });

    // Copy all files from source to destination
    await copyDirContents(sourceDir, releaseDir);

    console.log("Successfully copied favicon assets to release directory");
  } catch (error) {
    console.error("Error:", error instanceof Error ? error.message : error);
    process.exit(1);
  }
}

async function copyDirContents(src: string, dst: string): Promise<void> {
  // Ensure destination exists
  mkdirSync(dst, { recursive: true });

  // Read directory contents
  const entries = readdirSync(src);

  for (const entry of entries) {
    const srcPath = join(src, entry);
    const dstPath = join(dst, entry);
    const stats = statSync(srcPath);

    if (stats.isDirectory()) {
      // Recursively copy subdirectories
      await copyDirContents(srcPath, dstPath);
    } else {
      // Copy file
      copyFileSync(srcPath, dstPath);
      console.log(`Copied: ${srcPath} -> ${dstPath}`);
    }
  }
}

// Run the main function
if (import.meta.main) {
  main();
}

File: scripts/assets.rs
use std::fs;
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Define source and destination paths
    let source_dir = Path::new("assets/fav");
    let release_dir = Path::new("target/dx/hmziq-xyz/release/web/public/assets/fav");

    // Create the destination directory structure if it doesn't exist
    fs::create_dir_all(&release_dir)?;

    // Copy all files from source to destination
    copy_dir_contents(&source_dir, &release_dir)?;

    println!("Successfully copied favicon assets to release directory");
    Ok(())
}

fn copy_dir_contents(src: &Path, dst: &Path) -> Result<(), Box<dyn std::error::Error>> {
    // Ensure destination exists
    fs::create_dir_all(dst)?;

    // Iterate through source directory
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let path = entry.path();
        let file_name = entry.file_name();
        let dest_path = dst.join(&file_name);

        if path.is_dir() {
            // Recursively copy subdirectories
            copy_dir_contents(&path, &dest_path)?;
        } else {
            // Copy file
            fs::copy(&path, &dest_path)?;
            println!("Copied: {} -> {}", path.display(), dest_path.display());
        }
    }

    Ok(())
}
